### PR TITLE
[IOTDB-2039] Fix data redundant after too many open files exception occurs during compaction

### PR DESCRIPTION
**Pull request by THUMarkLau**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java [hunk 1]
```diff
@@ -233,4 +215,149 @@ public boolean checkValidAndSetMerging() {
...
+        targetTsFile.remove();
+        if (targetTsFile.getTsFile().exists()) {
+          LOGGER.warn(
+              "{} [Compaction][ExceptionHandler] failed to remove target file {}, set allowCompaction to false",
+              fullStorageGroupName,
+              targetTsFile);
+          tsFileManager.setAllowCompaction(false);
+          handleSuccess = false;
+        } else {
+          for (TsFileResource tsFileResource : selectedTsFileResourceList) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java:249]: add javadoc

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java [hunk 2]
```diff
@@ -30,7 +30,9 @@
 import org.apache.iotdb.db.engine.storagegroup.TsFileResourceList;
 import org.apache.iotdb.db.exception.WriteLockFailedException;
 import org.apache.iotdb.db.rescon.TsFileResourceManager;
+import org.apache.iotdb.tsfile.write.writer.RestorableTsFileIOWriter;
 
+import org.apache.commons.io.FileUtils;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java:35]: add read lock for being merged TsFile

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java [hunk 3]
```diff
@@ -233,4 +215,149 @@ public boolean checkValidAndSetMerging() {
...
+                new RestorableTsFileIOWriter(targetTsFile.getTsFile());
+            writer.close();
+            if (!writer.hasCrashed()) {
+              // target file is complete, delete source files
+              LOGGER.info(
+                  "{} [Compaction][ExceptionHandler] target file {} is complete, delete remaining source files",
+                  fullStorageGroupName,
+                  targetTsFile);
+              for (TsFileResource sourceFile : selectedTsFileResourceList) {
+                sourceFile.remove();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java:275]: add return bool

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java (hunk 1 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java (hunk 2 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java (hunk 3 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    if (targetTsFile.remove()) {
+      tsFileResourceList.writeLock();
+      try {
+        // add the source tsfile to TsFileResourceList
+        // in case of we have removed them from list before
+        for (TsFileResource tsFileResource : selectedTsFileResourceList) {
+          if (!tsFileResourceList.contains(tsFileResource)) {
+            tsFileResourceList.keepOrderInsert(tsFileResource);
+          }
+        }
```
[the remark this answers is not in the document]
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java:153]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java [hunk 5]
```diff
@@ -541,24 +542,64 @@ private static boolean isFileListHasModifications(
...
+          // we have to set modification offset to MAX_VALUE, as the offset of source chunk may
+          // change after compaction
+          modification.setFileOffset(Long.MAX_VALUE);
+          modificationFile.write(modification);
+        }
+      }
+    }
+  }
+
   public static void deleteTsFile(TsFileResource seqFile) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java:596]: could this be replaced by TsFileResource.remove()?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        targetTsFile);
+    if (!targetTsFile.remove()) {
+      // failed to remove target tsfile
+      // system should not carry out the subsequent compaction in case of data redundant
+      LOGGER.warn(
+          "{} [Compaction][ExceptionHandler] failed to remove target file {}",
+          fullStorageGroupName,
+          targetTsFile);
+      return false;
+    }
```
Comment (choubenson) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java:152]: .mods file of the targe file should also be deleted here.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+            handleWhenSomeSourceFilesLost(
+                fullStorageGroupName,
+                targetTsFile,
+                selectedTsFileResourceList,
+                tsFileResourceList,
+                lostSourceFiles);
+      }
+    }
+
+    if (handleSuccess) {
```
[the remark this answers is not in the document]
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java (hunk 6 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java:156]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    // deal with compaction modification
+    try {
+      for (TsFileResource sourceFile : selectedTsFileResourceList) {
+        if (sourceFile.getCompactionModFile().exists()) {
+          ModificationFile compactionModificationFile =
+              ModificationFile.getCompactionMods(sourceFile);
+          Collection<Modification> newModification = compactionModificationFile.getModifications();
+          compactionModificationFile.close();
+          // write the modifications to a new modification file
+          sourceFile.resetModFile();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java:166]: If mods exists, write to the existing mods file, not a new mods

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+      String fullStorageGroupName,
+      TsFileResource targetTsFile,
+      List<TsFileResource> selectedTsFileResourceList,
+      TsFileResourceList tsFileResourceList,
+      List<TsFileResource> lostSourceFiles) {
+    boolean handleSuccess = true;
+    try {
+      RestorableTsFileIOWriter writer = new RestorableTsFileIOWriter(targetTsFile.getTsFile());
+      writer.close();
+      if (!writer.hasCrashed()) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java:195]: Use a method such as:   TsFileUtils.isTsFileComplete() to replace this writer

**Pull request merged**